### PR TITLE
change compiler extension to work as part of an "environment"

### PR DESCRIPTION
**Pull request by davidfirst**

### CODE REVIEW

File: src/extensions/environments/environment.ts [hunk 1]
```diff
@@ -1,5 +1,10 @@
+import { Tester } from './../tester';
+import { Compiler } from '../compile';
```
Comment (GiladShoham) [inline src/extensions/environments/environment.ts:2]: @davidfirst This make a cyclic deps between evns and compiler/tester
we should extract the interfaces to the general types I think (or use import type at least once we support it)

File: src/extensions/typescript/typescript.compiler.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    return { taskFile: 'transpile' };
+  }
+  compileFile(
+    fileContent: string,
+    options: { componentDir: string; filePath: string }
+  ): { outputText: string; outputPath: string }[] | null {
+    const supportedExtensions = ['.ts', '.tsx'];
+    const fileExtension = path.extname(options.filePath);
+    if (!supportedExtensions.includes(fileExtension)) {
+      return null; // file is not supported
```
Comment (GiladShoham) [inline src/extensions/typescript/typescript.compiler.ts:18]: change to undefined

File: src/extensions/typescript/typescript.compiler.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    }
+    const compilerOptionsFromTsconfig = ts.convertCompilerOptionsFromJson(this.tsConfig.compilerOptions, '.');
+    if (compilerOptionsFromTsconfig.errors.length) {
+      throw new Error(`failed parsing the tsconfig.json.\n${compilerOptionsFromTsconfig.errors.join('\n')}`);
```
Comment (GiladShoham) [inline src/extensions/typescript/typescript.compiler.ts:22]: make an error class for this

File: src/extensions/typescript/typescript.compiler.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        getCanonicalFileName: p => p,
+        getCurrentDirectory: ts.sys.getCurrentDirectory,
+        getNewLine: () => ts.sys.newLine
+      };
+      const error = ts.formatDiagnosticsWithColorAndContext(result.diagnostics, formatHost);
+
+      throw new Error(error);
+    }
+
+    const replaceExtToJs = filePath => filePath.replace(new RegExp(`${fileExtension}$`), '.js'); // makes sure it's the last occurrence
```
Comment (GiladShoham) [inline src/extensions/typescript/typescript.compiler.ts:43]: move to utils
Also, see - get-ext in the old utils/fs to get the last one.
You can also use getWithoutExt (inside the fs-no-ext file).

File: src/extensions/typescript/typescript.compiler.ts (hunk 2 above)
Comment (davidfirst) [reply src/extensions/typescript/typescript.compiler.ts:18]: @GiladShoham why `undefined`? it is defined and it is null.

File: src/extensions/environments/environment.ts (hunk 1 above)
Comment (davidfirst) [reply src/extensions/environments/environment.ts:2]: @GiladShoham , yeah, I was thinking about it. But it's only for the type. Once we migrate "prettier" to v2, we can update Babel to the latest version that supports `import type`. Then, you won't see any circular.

**Pull request merged**